Sort #includes

OnlineCop opened the pull request:
1. Header files should never use 'using ...' to avoid contaminating the namespace of the `#include`-ing file.
2. It's fine to use 'using ...' in the .cpp files, although in this PR, I opted to prepend `std::` on all STL types until the refactor is complete.
3. There are a LOT of changes, but I tried to break these up so each commit does only one type of change each.
4. PR reviews have been pretty slow lately, so I'll leave this here for a little while to see if it garners any feedback or compilation errors from others. Then in it goes!
5. The order that headers should be included in the .cpp files are as follows:
    1. Source file's counterpart: `disk.cpp` should `#include "disk.h"` first.
    2. Project files, sorted alphabetically: all `#include "..."` should come before `#include <...>`.
    3. System files, sorted alphabetically: all `#include <...>` files.

If there must be an exception to rule 5, please clearly comment it (example: win.cpp needs a `#define` before a specific Windows header, before any other `#include ...` file).

OnlineCop committed:
CMakeLists.txt: Add tmx_animation.cpp, tmx_tileset.cpp

Moves class definitions from .h to .cpp.

OnlineCop committed:
Change func_name(void) => func_name()

OnlineCop committed:
Update Copyright info

Adds the same copyright to the top of all .cpp and .h files, except include/gettext.h.

OnlineCop committed:
Prefix STL types with std:: in .h files

OnlineCop committed:
Comment change: From same-line to next line, no functional changes

Changes from one-line:
    'type var; /*!< desc */'
to two-line:
    /*! desc */
    type var;

OnlineCop committed:
Remove 'using std:: ...' from .h files (WARNING: compile errors)

It's fine to have 'using ...' in the .cpp files; not okay to litter any .cpp file that #includes those files, though.

OnlineCop committed:
Get rid of 'using std::'

As part of a larger refactoring, I'm going to temporarily remove 'using std::...' and prepend 'std::' to all the STL types.

A future commit will further remove any kind of 'using ...' in all .h files; they should be okay to remain in .cpp files.

OnlineCop committed:
Move 'using tinyxml2:...' and 'using namespace tinyxml2' from .h to .cpp

This prepends 'tinyxml2::' in the .h files, but leaves them off in the .cpp files.
- Basically, we never want to pollute the namespace when including a file that may exist in a different namespace.

OnlineCop committed:
Fix '#include <itemdefs.h>' => '#include "itemdefs.h"'

OnlineCop committed:
Sort #includes

OnlineCop committed:
Remove a bunch of unnecessary #includes

OnlineCop committed:
Remove even more #includes

SDL.h is the most notable in this commit: I want as many #includes removed from .h as possible, so they have to be included in the .cpp files now.

pedro-w commented:
> PR reviews have been pretty slow lately

I can't keep up. Unfortunately I haven't got as much time for this (or any other project) as I'd like

> The order that headers should be included

I've been using `clang-format` before committing* and that has its own idea about header order - I think it does put XYZ.h at the top in file XYZ.cc, then system headers then our headers but also respects a blank line to mark groups of headers that won't be mixed. 
I recommend we should use clang-format because it's very good at avoiding bogus white-space changes in diffs.

\* Unless I forget...

pedro-w commented on include/compat.h:
```diff
@@ -1,3 +1,24 @@
+       675 Mass Ave, Cambridge, MA 02139, USA.
+*/
+
 #pragma once
```
Would be good to eliminate this file altogether - it was supposed to be some quick definitions to paper over the cracks between Allegro 4 and SDL2

OnlineCop replied:
I agree. I scripted out something to first remove the header from all files, then re-insert a standard one into all of them.

Although, I would probably prefer to just remove it from all files and have a LICENSE.md or something in the project root. Then there only ever needs to be 1 change (like updating the Copyright year).

I think we also planned to remove the outdated source files outside of the `src/` and `include/` directories, as those were very Allegro- and C-specific.

pedro-w replied:
I'd leave them, but just remove the \\page directive from all but one (on kq.cpp.) I thought for a moment we didn't actually include the licence anywhere but (as the header says!) it's in the file COPYING. I don't know what we do about our dependency libraries - it is probably enough to reference them at the end of the README file (just say KQ depends on XXX library copyright YYY and give the project URL)

pedro-w commented on include/maps.h:
```diff
@@ -60,28 +58,50 @@ enum eMapMode
-    string song_file;   /*!< Base file name for map song */
-    string map_desc;    /*!< Map name (shown when map first appears) */
-    KMarkers markers;   /*!< Marker array */
-    KBounds bounds;     /*!< Bound array and bound size */
```
Doesn't really matter but I found the 'point-left' comment format easier to read

OnlineCop replied:
I plan to re-add those 'point-left' comments. I've got a lot of reformatting, and I've been using a bunch of regex to match everything and didn't want to accidentally leave the comments off.

pedro-w commented on include/menu.h:
```diff
@@ -47,37 +45,39 @@ struct KQuestItem
+    void display_quest_window();
     void status_screen(size_t);
-    vector<KQuestItem> quest_list;
+    std::vector<KQuestItem> quest_list;
```
Good grief, 22 / 83 files viewed, I didn't know there even was that number of files in KQ

pedro-w commented on include/random.h:
```diff
@@ -1,4 +1,4 @@
-/*
+/*! \page License
    KQ is Copyright (C) 2002 by Josh Bolduc
```
Actually I think we only want one of these with the Doxygen markup. It'll create a standalone page called License in the generated documents, so there only needs to be one in the whole project, not one per file (maybe Doxygen is smart enough to merge them, I haven't checked)

OnlineCop replied:
That would be great!

pedro-w replied:
I checked, on Doxygen 1.9.3 at least, it seems that if you have identical \\page identifiers with identical content you'll get one page with that content. If the content differs it will concatenate one of each of all the different versions. So, smarter than I gave it credit for!

OnlineCop replied:
What are your thoughts? Remove from all files except maybe `kq.cpp` (since it includes `int main()`)? Keep them all in for now (maybe remove them in a separate PR)? Remove from everything and just write a LICENSE.md file?

Which, by the way, we probably need anyway if this is going to be using `gettext`, `tinyxml2` and others: we may need to include all those licenses somewhere anyway so it will be valid on the various Linux distributions.

pedro-w replied:
I'd leave them, but just remove the \page directive from all but one (on kq.cpp.) I thought for a moment we didn't actually include the licence anywhere but (as the header says!) it's in the file COPYING. I don't know what we do about our dependency libraries - it is probably enough to reference them at the end of the README file (just say KQ depends on XXX library copyright YYY and give the project URL)

pedro-w commented on src/combat.cpp:
```diff
@@ -27,36 +27,20 @@
-#include <iostream>
-
 #include "combat.h"
-#include "constants.h"
```
Halfway... keep going!
Removing these unnecessary includes is very useful, cuts down the incremental recompilation time quite a bit

OnlineCop replied:
That's one of my motivations for doing that. Like if anyone changes `kq.h`. Recompiles nearly EVERYTHING.

pedro-w commented on src/magic.cpp:
```diff
@@ -31,14 +30,15 @@
-#include "ssprites.h"
 #include "structs.h"
 
+#include <cstdio>
```
I don't think cstdio and cstring are used for anything in this file?

pedro-w commented on src/masmenu.cpp:
```diff
@@ -29,12 +28,15 @@
 #include "selector.h"
 #include "setup.h"
+
+#include <cstdio>
```
cstdio and cstring not used AFAICS - can be removed?

pedro-w commented on src/sgame.cpp:
```diff
@@ -55,11 +52,15 @@
 #include "structs.h"
 #include "timing.h"
 
+#include <cctype>
```
Final push 72/83 ... I don't think cctype, cstdio or cstring are needed

pedro-w commented on src/tmx_tileset.cpp:
```diff
@@ -0,0 +1,81 @@
+/*! \page License
```
Do the default constructors and assignment operators not do what we want here? (ditto tmx_animation.cpp)

OnlineCop replied:
C++ has a weird history with when the compiler will generate the constructors and assignment operators for copy/move. I keep forgetting when the defaults are generated, or when it acts as though you had explicitly `= delete`d them.

For KTmxAnimation, I just directly used `= default` since those are all trivial, but I didn't want one of them to accidentally get `= delete`d because I forgot one or more of them.

For KTmxTileset, it does a shallow copy, and I don't know how we wanted to treat `Raster* imagedata`. Do we want two objects to point to the same pointer? Does the tmx_tileset only point to it as a reference, or does it own the memory it points to?

pedro-w replied:
The `Raster*` is a non-owning pointer, the data is owned by the image cache in imgcache.cpp

pedro-w reviewed (commented):
Looks fine, I left some comments. Overall seems a lot of changes, hopefully won't see too many more of these  😄 💦

OnlineCop commented:
> > PR reviews have been pretty slow lately
> 
> I can't keep up. Unfortunately I haven't got as much time for this (or any other project) as I'd like

No worries; it seems like you've been the only one reviewing stuff; where did @z9484 and everyone else go?

In the very rare hours/days that I actually get enough time to sit down to this project, I'm hoping to rapid-fire a bunch of refactors to the engine. My top priority isn't so much the story line (I really hope others can work on that), but rather to push as much _out_ of the game engine as possible and into the scripts.

I'd love it if KQtheBetrayor could be revived one day as a result of these changes; I think I played it once (maybe not all the way through) but that was around the time that I had so many problems getting Allegro to work with newer systems so I could actually see much of it.

@teamterradactyl If you have any free time, do you want to help with the story elements for the 25th anniversary? Any help at this point is appreciated (but not required).

OnlineCop commented:
> I've been using `clang-format` before committing* and that has its own idea about header order - I think it does put XYZ.h at the top in file XYZ.cc, then system headers then our headers but also respects a blank line to mark groups of headers that won't be mixed.
> I recommend we should use clang-format because it's very good at avoiding bogus white-space changes in diffs.

🤦‍♂️ 
I 100% forgot about clang-format. It's not installed natively on my Windows system BUT I'm able to clean up all these files via WSL.

I feel very dumb for not using this from the get-go. And then when I was looking at the after-effects of running clang-format on all these files and thought, "Huh: this was set up almost exactly like _I'd_ have preferred the code to be formatted" and decided to `git blame .clang-format`. It was me. I created that file. 2 years ago.

Derp.

See what I keep saying about forgetting stuff? This. This is one of those times.

Update to the PR will be ready soon. :shaking-my-head:

OnlineCop committed:
Passed through clang-format

pedro-w committed:
Merge branch 'road_to_1.0' into dev/onlinecop/sort_includes

pedro-w commented:
> 🤦‍♂️ I 100% forgot about clang-format. It's not installed natively on my Windows system BUT I'm able to clean up all these files via WSL.

I had a feeling I'd used clang-format directly from Visual Studio - the following link suggests support is 'built in'?

https://devblogs.microsoft.com/cppblog/clangformat-support-in-visual-studio-2017-15-7-preview-1/

OnlineCop committed:
All files except kq.cpp/kq.h: Remove '\page License' from copyright

Obviously, this excludes include/gettext.h, which is a 3rd-party convenience header.

OnlineCop commented:
Unless there are other changes anyone feels _must_ go into this PR, I'm hoping to call it good for now, so I can start work on a few other refactors, many of which were relying on a lot of the changes made here.

Those changes should primarily be for bugs this PR may have introduced (I may have removed too many headers and, while it compiles on _my_ system, may have been broken on _yours_; that kind of thing). If you spot more refactoring/cleaning that should occur, create separate Issues so they can be tracked and closed.

pedro-w commented:
It didn't compile out of the box on Mac so I made PR #108

pedro-w commented:
> I'm able to clean up all these files via WSL

Just as a matter of interest what version of clang-format have you? Because sometimes I  get some changes from running it on code that you've already run it on and checked in. I've got "clang-format version 14.0.0"

OnlineCop commented:
The latest that Ubuntu via Windows WSL would pull with `apt install` was something like 10.0 or 12.0.

I ended up just installing LLVM 14 so it would have clang-format-14, which may be what others will end up having to do if they want to use it.

Occasionally, I'd run clang-format and not like the results in certain areas, so I'd revert those changes before pushing. I don't mind if we want to blanket "format everything" or if you want to pick-and-choose things that would just look cleaner by reverting small sections.

pedro-w commented:
Please please please can we not do any more of these general all-over kind of commits. I had a couple of bug fixes which I had been working on but they no longer apply to the current road_to_1.0 branch and it has taken me literally an hour to try and resolve the conflicts over and over again in a git rebase. And I've almost certainly made some mistakes. I know it's good to be tidy and consistent but things like renaming variables surely can't be our priority. I say this because I expect I shall have to go through the pain again with #116

pedro-w commented:
Sorry for that rant. I'm not blaming anyone but myself (and maybe Linus Torvalds 😉 ) for my git woes. Usually git rebase works like a charm but occasionally not. Anyway I guess I'll just have to do a hard reset and try again...
